### DISCUSSION TITLE
When `z.infer` a `Record` with `Brand` as a key, it becomes a Partial Record.

### DISCUSSION THREAD

**Discussion by Hiroshiba** [colinhacks/zod#2069] [Q&A] [answered]
When `z.infer` a `Record` with `Brand` as a key, it becomes a Partial Record.

With defining a Record type normally:
```ts
import { z } from "zod";
const zodKey = z.string().brand("zodKey");
type ZodKey = z.infer<typeof zodKey>;

type ZodRecord = Record<ZodKey, number>;
// type ZodRecord = {
//   [x: string & z.BRAND<"zodKey">]: number;
// }
```

With defining a Record type with `z.infer`:
```ts
import { z } from "zod";
const zodKey = z.string().brand("zodKey");
const zodRecord = z.record(zodKey, z.number());

type ZodRecord = z.infer<typeof zodRecord>;
// type ZodRecord = {
//   [x: string & z.BRAND<"zodKey">]: number | undefined;
// }
```

It has `undefined`.

**Comment by JacobWeisenburger** [comment]
Technically in TypeScript, when you use `string` as the key type for `Record`, you have a partial record, because you can have no properties at all, and there are no errors. TypeScript is just being implicit about it.

```ts
const partialRecord: Record<string, string> = {} // no errors
```

I'm not sure why the `undefined` type shows up when using a branded string as the key. But it doesn't seem to change the behavior as far as I can see.

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by Hiroshiba** [reply]
Thank you!
The somewhat tricky part is that the behavior is different from normal TypeScript, so I get an error with the assignment.

```ts
const zodKey = z.string().brand("zodKey");

// with TypeScript
type ZodKey = z.infer<typeof zodKey>;
type ZodRecord1 = Record<ZodKey, number>;

// with Zod (will be Partial)
const zodRecord = z.record(zodKey, z.number());
type ZodRecord2 = z.infer<typeof zodRecord>;

const record2: ZodRecord2 = {};
const record1: ZodRecord1 = record2; // error
```

I can make `ZodRecord2` also `Partial` or use `as` to work around this, I'd like to know if there is a better solution.

**Comment by Hiroshiba** [comment]
I followed the discussion.

Here `Record` type is now `Partial`.
https://github.com/colinhacks/zod/issues/751

Here `key=string` is no longer `Partial`.
https://github.com/colinhacks/zod/issues/992#issuecomment-1057290762

I thought it was more intuitive to not make it a `Partial` type to match vanilla TypeScript.

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const zodKey = z.string().brand( "zodKey" )

// with TypeScript
type ZodKey = z.infer<typeof zodKey>
type ZodRecord1 = Record<ZodKey, number | undefined>

// with Zod (will be Partial)
const zodRecord = z.record( zodKey, z.number() )
type ZodRecord2 = z.infer<typeof zodRecord>

const record2: ZodRecord2 = {}
const record1: ZodRecord1 = record2 // no error, yay!
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by Hiroshiba** [reply]
What I want is for the elements of the `Record` not to be `undefined`.
Of course, I know this is not type safe.

I think the philosophy of `zod` is that certainty is more important than conformance to TypeScript, so it is designed to make the `Record` be `Partial`.
But well, I think it is a little inconvenient for the user because there are two different philosophies.

Either use `as` and adapt to TypeScript, or make it `Partial` and adapt to `zod`, I'll give this some more thought.

**Comment by JacobWeisenburger** [comment, marked as answer]
It seems the problem is caused by this code:

https://github.com/colinhacks/zod/blob/981af6503ee1be530fe525ac77ba95e1904ce24a/src/types.ts#L3061-L3069

If it were changed to: `[K] extends [string]`, it might work for you.
However, I'm not sure if that will be a breaking change for other users.

**Reply by Hiroshiba** [reply]
Thank you for the code!!!
How about making `Record<K, V>` even if `K` is `z.BRAND`?

So, the code would be something like this.
```ts
type BRANDED_TYPE = "???";
type RecordType<K extends string | number | symbol, V> = [string] extends [K]
  ? Record<K, V>
  : [number] extends [K]
  ? Record<K, V>
  : [symbol] extends [K]
  ? Record<K, V>
  : [BRANDED_TYPE] extends [K] // changed point
  ? Record<K, V>
  : Partial<Record<K, V>>;
```

Sorry, I'm not familiar with TypeScript and don't know what to write for `BRANDED_TYPE`...

**Reply by Hiroshiba** [reply]
How about like this...?

```ts
type RecordType<K extends string | number | symbol, V> = [string] extends [K]
  ? Record<K, V>
  : [number] extends [K]
  ? Record<K, V>
  : [symbol] extends [K]
  ? Record<K, V>
  : K extends z.BRAND<infer B> // changed point
  ? Record<K, V>
  : Partial<Record<K, V>>;
```

**Reply by JacobWeisenburger** [reply]
that could potentially work. I'm not sure about the unintended consequences, but it is better than what I suggested. Perhaps you could submit a PR for this?

**Reply by Hiroshiba** [reply]
OK, I'll give it a try!

**Reply by Hiroshiba** [reply]
Done! https://github.com/colinhacks/zod/pull/2097

**Reply by googol** [reply]
For me this doesn't seem to fix the problem.

It seems like I can only get it to work when I replace the zod RecordType with the TS built in record. What is the reasoning behind this type anyways? I see there's a comment about the incompleteness of the TS record, but that's handled by the strictness option [`noUncheckedIndexedAccess `](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess).

To me it seems removing this type and replacing it with the plain `Record` is the best option

**Reply by EvilAlexei** [reply]
Same for me, this doesn't solve the problem.
I tried to use an `enum` or `union` as the key, but my record became `Partial`,
it seems a bit strange.

**Reply by googol** [reply]
I have created a pull request to simplify the `z.record` like I described: https://github.com/colinhacks/zod/pull/2287

**Reply by wdanilo** [reply]
It doesn't seem to work for me. If I create:

```ts
function brandedStringCons<T extends symbol>() {
  const Schema = Z.string().brand<T>()
  type Out = Z.infer<typeof Schema>
  const cons = (val: string): Out => val as Out
  cons.Schema = Schema
  return cons
}

declare const IdSymbol: unique symbol
export const Id = brandedStringCons<typeof IdSymbol>()
export type Id = ReturnType<typeof Id>
```

And then use `Z.record(Id.Schema, number)`, then the inferred type will be `Partial<Record<...>>`

**Comment by fwoelffel** [comment]
I really can't get my head around the `RecordType` definition:
```ts
export declare type RecordType<K extends string | number | symbol, V> = [
    string
] extends [K] ? Record<K, V> : [number] extends [K] ? Record<K, V> : [symbol] extends [K] ? Record<K, V> : [BRAND<string | number | symbol>] extends [K] ? Record<K, V> : Partial<Record<K, V>>;
```

Shouldn't it rather be
```ts
export declare type RecordType<K extends string | number | symbol, V> = [
    K
] extends [string] ? Record<K, V> : [K] extends [number] ? Record<K, V> : [K] extends [symbol] ? Record<K, V> : [K] extends [BRAND<string | number | symbol>] ? Record<K, V> : Partial<Record<K, V>>;
```

WDYT? I also don't quite get how we could have a `Partial<Record<K, V>>` when `K extends string | number | symbol` and all those cases are covered in the type ternary? Am I mistaken to think it should rather be `never`?

**Reply by PrzemoRevolve** [reply]
We're also experiencing this issue:

```ts
import { z } from 'zod';

const myKeySchema = z.coerce.number().brand('myKey');
const myRecordSchema = z.record(myKeySchema, z.number());
type MyRecord = z.infer<typeof myRecordSchema>;

const myRecord: MyRecord = {};
```

`myRecord` is inferred as `Partial<Record<number & z.BRAND<"myKey">, number>>`

**Reply by jthemphill** [reply]
> Shouldn't it rather be...

@colinhacks explains why this would not be correct in https://github.com/colinhacks/zod/pull/2287#issuecomment-2013497852

Zod's type definition is trying to say that the record is partial when the key is an enum or literal.

If the key is a literal, like `K = "some string"`, then `K extends string` is true, but `string extends K` is not true.

**Comment by jthemphill** [comment]
I believe I have a correct fix for this issue, which I'd like to submit for Zod v4:

* https://github.com/colinhacks/zod/pull/4277

With this PR, we are still returning a partial record if the key is a literal or enum. This means the typing of `ZodRecord` will still align with the justification given in https://github.com/colinhacks/zod/pull/2287#issuecomment-2013497852.

But now, if the key is a branded string/number/symbol, then the record will no longer be turned into a `Partial<Record<K, V>>`. A branded string will be handled as a string.
